File: ios/UnityShowView.swift
import ExpoModulesCore
import UIKit

internal final class UnityShowView: ExpoView {
  private var source: [String: Any?]?
  private var paused = false
  private var listenerBox: UnityShowEventListenerBox?

  private let onUnityReady = EventDispatcher()
  private let onUnityMessage = EventDispatcher()
  private let onUnityStateChange = EventDispatcher()
  private let onUnityError = EventDispatcher()

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    backgroundColor = .clear
    listenerBox = UnityShowEventBus.addListener { [weak self] eventName, body in
      self?.dispatchEvent(eventName, body: body)
    }
  }

  deinit {
    UnityShowEventBus.removeListener(listenerBox)
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()

    if window == nil {
      UnityShowRuntime.shared.detach(from: self)
      return
    }

    attachUnityIfReady()
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    subviews.forEach { view in
      view.frame = bounds
    }
  }

  func setSource(_ source: [String: Any?]?) {
    self.source = source
    attachUnityIfReady()
  }

  func setPaused(_ paused: Bool) {
    if self.paused == paused {
      return
    }

    self.paused = paused

    do {
      if paused {
        try UnityShowRuntime.shared.pause()
      } else {
        try UnityShowRuntime.shared.resume()
      }
    } catch {
      dispatchError(
        code: "ERR_UNITY_LIFECYCLE",
        message: error.localizedDescription,
        recoverable: true
      )
    }
  }

  private func attachUnityIfReady() {
    guard window != nil else {
      return
    }

    do {
      try UnityShowRuntime.shared.attach(to: self, source: source)
    } catch {
      dispatchError(
        code: "ERR_UNITY_FRAMEWORK_UNAVAILABLE",
        message: error.localizedDescription,
        recoverable: true
      )
    }
  }

  private func dispatchEvent(_ eventName: String, body: [String: Any?]) {
    switch eventName {
    case "ready":
      onUnityReady(body as [String: Any])
    case "message":
      onUnityMessage(body as [String: Any])
    case "stateChange":
      onUnityStateChange(body as [String: Any])
    case "error":
      onUnityError(body as [String: Any])
    default:
      break
    }
  }

  private func dispatchError(code: String, message: String, recoverable: Bool) {
    onUnityError([
      "code": code,
      "message": message,
      "recoverable": recoverable
    ])
  }
}

File: android/src/main/java/com/reactnativeunityshow/UnityShowView.kt
package com.reactnativeunityshow

import android.annotation.SuppressLint
import android.content.Context
import android.view.ViewGroup
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView

@SuppressLint("ViewConstructor")
internal class UnityShowView(context: Context, appContext: AppContext) :
  ExpoView(context, appContext) {
  override val shouldUseAndroidLayout = true

  private var source: Map<String, Any?>? = null
  private var paused = false
  private var listening = false
  private val eventListener: UnityShowEventListener = { eventName, body ->
    post {
      dispatchEvent(eventName, body)
    }
  }

  private val onUnityReady by EventDispatcher<Map<String, Any?>>()
  private val onUnityMessage by EventDispatcher<Map<String, Any?>>()
  private val onUnityStateChange by EventDispatcher<Map<String, Any?>>()
  private val onUnityError by EventDispatcher<Map<String, Any?>>()

  init {
    layoutParams = ViewGroup.LayoutParams(
      ViewGroup.LayoutParams.MATCH_PARENT,
      ViewGroup.LayoutParams.MATCH_PARENT
    )
  }

  fun setSource(source: Map<String, Any?>?) {
    this.source = source
    attachUnityIfReady()
  }

  fun setPaused(paused: Boolean) {
    if (this.paused == paused) {
      return
    }

    this.paused = paused

    if (!isAttachedToWindow) {
      return
    }

    try {
      if (paused) {
        UnityShowRuntime.pause()
      } else {
        UnityShowRuntime.resume()
      }
    } catch (exception: Exception) {
      dispatchError("ERR_UNITY_LIFECYCLE", exception.message, true)
    }
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    startListening()
    attachUnityIfReady()
  }

  override fun onDetachedFromWindow() {
    stopListening()
    UnityShowRuntime.detachFrom(this)
    super.onDetachedFromWindow()
  }

  fun onViewDestroys() {
    stopListening()
    UnityShowRuntime.detachFrom(this)
    removeAllViews()
  }

  private fun attachUnityIfReady() {
    if (!isAttachedToWindow) {
      return
    }

    val activity = appContext.currentActivity
    if (activity == null) {
      dispatchError(
        "ERR_UNITY_ACTIVITY_UNAVAILABLE",
        "The current Android activity is required to attach Unity.",
        true
      )
      return
    }

    try {
      UnityShowRuntime.attachTo(this, activity, source)
    } catch (exception: Exception) {
      dispatchError("ERR_UNITY_RUNTIME_UNAVAILABLE", exception.message, true)
    }
  }

  private fun startListening() {
    if (listening) {
      return
    }

    UnityShowEventBus.addListener(eventListener)
    listening = true
  }

  private fun stopListening() {
    if (!listening) {
      return
    }

    UnityShowEventBus.removeListener(eventListener)
    listening = false
  }

  private fun dispatchEvent(eventName: String, body: Map<String, Any?>) {
    when (eventName) {
      "ready" -> onUnityReady(body)
      "message" -> onUnityMessage(body)
      "stateChange" -> onUnityStateChange(body)
      "error" -> onUnityError(body)
    }
  }

  private fun dispatchError(code: String, message: String?, recoverable: Boolean) {
    onUnityError(
      mapOf(
        "code" to code,
        "message" to (message ?: "Unity Android integration failed."),
        "recoverable" to recoverable,
      )
    )
  }
}
